CDCgov/prime-reportstream
# Experience/2245: Populate Patient Address Fields with Ordering Facility Address when Patient Address is Blank

This PR introduces the IfNotPresentMapper, which checks if one or more elements are blank or not present on a row, and if so, will replace an element's value with either some literal string value or the value from a different field on a row
ex. 
mapper: ifNotPresent($mode:literal, $string:NO ADDRESS, patient_zip_code, patient_state)
- if patient_zip_code and patient_state are missing or blank, then replace element's value with "NO ADDRESS"
mapper: ifNotPresent($mode:lookup, ordering_provider_city, patient_zip_code)
- if patient_zip_code is missing or blank, then replace element's value with that of the ordering_provider_city

**If you are suggesting a fix for a currently exploitable issue, please disclose the issue to the prime-reportstream team directly outside of GitHub instead of filing a PR, so we may immediately patch the affected systems before a disclosure. See [SECURITY.md/Reporting a Vulnerability](https://github.com/CDCgov/prime-reportstream/blob/master/SECURITY.md#reporting-a-vulnerability) for more information.**

Test Steps:
1. Include the new mapper in a schema element
ex. 	  
```
- name: patient_street
  csvFields: [{ name: patientHomeAddress_pii}]
  mapper: ifNotPresent($mode:literal, $string:*** No Address Given ***, patient_zip_code, patient_state)
  - name: patient_city
  csvFields: [{ name: patientCity_pii}]
  mapper: ifNotPresent($mode:lookup, ordering_provider_city, patient_zip_code, patient_state)
```
2. Update the data so that your "conditional fields" are blank (i.e. patient_zip_code and patient_state from above ex.)
3. Send the data to RS using that schema
4. Verify that your schema element's value was either with your literal (i.e. *** No Address Given *** from above ex.) or lookup value (i.e. value of "ordering_provider_city" from above ex.)

## Changes
- added `$mode:` mapper token to take a "literal" or "lookup" value to determine the type of operation of IfNotPresent
- added `$string:` mapper token to indicate the type of value the "literal" mode is using (currently just String...)

## Checklist

### Testing

### Process

## Fixes
- #issue - List GitHub issues this PR fixes

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
